할인 목록을 조회하라

hannut91 opened the pull request:
저장되어있는 할인 목록을 반환하는 getDiscounts를 만들었습니다.
할인 목록을 조회할 수 있도록 DiscountController를 만들었습니다. `GET /discounts`로 요청하면, 할인 목록을 응답합니다.

hannut91 committed:
Handle request discount list

할인 목록을 조회할 수 있도록 DiscountController를 만들었습니다.
`GET /discounts`로 요청하면, 할인 목록을 응답합니다.

hannut91 committed:
Get discounts in Discount repository

저장되어있는 할인 목록을 반환하는 getDiscounts를 만들었습니다.

etff reviewed (changes requested):
고생하셨습니다. 생각해볼만한 코멘트를 남겼는데 확인부탁드립니다.

hannut91 committed:
객체 맵퍼 mapstruct를 의존성에 추가하라

엔티티 객체와 응답을 위한 DTO의 매핑을 쉽게하기 위해서
mapstruct라이브러리를 설치했습니다.

io.springfox:springfox-boot-starter:3.0.0에서 이미 mapstruct 구버전에
의존하고 있어, 여기서 의존하고 있는 mapstruct는 설치하지 않도록 하고
따로 명시한 mapstruct를 설치하도록 했습니다.

See also
  - https://mapstruct.org/

hannut91 committed:
할인 목록을 담고 있는 일급객체를 생성하라

할인 목록을 담고 있는 일급객체인 DiscountList를 만들고, 응답을 위한
DiscountData을 매핑해줄 수 있도록 DiscountDataMapper를 만들었습니다.
DiscountList는 list()메서드를 통해 매핑된 객체 목록을 반환합니다.

hannut91 committed:
엔티티를 직접 응답하는 대신, 할인 목록을 응답하라

할인 엔티티를 직접 노출하게 되면, 엔티티에 반영된 결과가 응답 결과로
바로 반영이 되므로, 응답이 어떻게 변화될지 예측하기 어렵습니다.
따라서 할인 엔티티 응답을 위한 DiscountData목록을 응답하도록 했습니다.

etff commented on src/main/java/com/kurly/promotion/ui/DiscountData.java:
```diff
@@ -0,0 +1,18 @@
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
```
@Data 를 써보면 어떨까요?

hannut91 replied:
적용했습니다 감사합니다 ㅎㅎ

etff commented on src/main/java/com/kurly/promotion/ui/DiscountList.java:
```diff
@@ -0,0 +1,19 @@
+
+    public DiscountList(List<Discount> discountList) {
+        this.discountList = discountList;
+    }
```
DiscountList는 도메인 객체일까요?

hannut91 replied:
도메인 객체는 아니고, 응답 객체 생성을 위한 객체인데 무엇이 잘못되었을까요?

etff replied:
도메인의 1급 컬렉션을 발견하면 도메인 내에서 무언가를 한다고 기대를 하게됩니다.

하지만 구현하신 내용을 보면 DTO를 만들어내는 행위를 하고 있습니다. 
도메인에서 DTO를 만드는 건 도메인의 역할이 아니기 때문에 코멘트를 달았습니다. 

혹시 추가적인 대화가 필요하시면 DM 부탁드립니다.

etff commented on src/test/java/com/kurly/promotion/ui/DiscountListTest.java:
```diff
@@ -0,0 +1,35 @@
+                    .isEqualTo(discountList.get(0).getProductId());
+        }
+    }
+}
```
파일 끝은 공백을 생성해주세요

etff commented on src/main/java/com/kurly/promotion/ui/DiscountList.java:
```diff
@@ -0,0 +1,20 @@
+        return this.discountList.stream()
+                .map(DiscountDataMapper.INSTANCE::toDto)
+                .collect(Collectors.toList());
+    }
```
이 클래스가 도메인이라면 도메인이 DTO에 의존하고 있습니다.

굳이 클래스를 생성하는 것보다 DTO에서 변환해보면 어떨까요?

etff commented on src/main/java/com/kurly/promotion/ui/DiscountList.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.stream.Collectors;
+
+public class DiscountList {
+    private final List<Discount> discountList;
```
클래스나 메서드 이름에 자료형을 넣는 것보다 복수형으로 표현해보면 어떨까요?

etff commented on src/main/java/com/kurly/promotion/ui/DiscountDataMapper.java:
```diff
@@ -0,0 +1,12 @@
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+@Mapper
+public interface DiscountDataMapper {
+    DiscountDataMapper INSTANCE = Mappers.getMapper(DiscountDataMapper.class);
+
+    DiscountData toDto(Discount discount);
+}
```
메퍼 활용 좋습니다 👍

etff commented on src/main/java/com/kurly/promotion/ui/DiscountData.java:
```diff
@@ -0,0 +1,18 @@
+
+    Integer flatRate;
+
+    Long productId;
```
접근 제어자는 없어도 되는걸까요?

hannut91 replied:
private으로 명시해야겠습니다

etff reviewed (changes requested):
고생하셨습니다! 생각해볼만한 코멘트를 남겼습니다

etff commented on src/main/java/com/kurly/promotion/ui/DiscountController.java:
```diff
@@ -0,0 +1,28 @@
+        DiscountList discountList = new DiscountList(
+                discountService.getDiscounts()
+        );
+        return discountList.list();
```
해당 메서드가 무엇을 하는지 예측하기 힘듭니다. 
동사를 사용해서 어떤 행위를 하는지 표현해보면 어떨까요?

hannut91 replied:
이 메서드가 그냥 `list내놔` 라고 하는거라서, 제가보기에는 괜찮아 보이는데 다른 좋은 이름을 제안해주실 수 있나요?

etff replied:
메서드는 동작/행위(Behavior)를 기대하는데 `list()`는 행위라 보기 힘들지않을까요?
리스트 요청이면 `getList()` , List로 변환을 기대한다면 `toList()`를 쓰면 좋을 것같습니다.

hannut91 committed:
분리된 컨트롤러를 하나로 합쳐라

종명님이 만드신 `com.kurly.promotion.controller`와
`com.kurly.promotion.ui`가 두개로 나뉘어져서 `com.kurly.promotion.ui`로
합쳤습니다.

etff reviewed (approved):
고생하셨습니다

hannut91 committed:
DiscountList를 사용하지 않고, DTO로 변환하여 응답하라

DiscountList는 도메인의 일급컬렉션처럼 생겨서, 헷갈릴 수 있어서
사용하지 않기로 하였습니다.

hannut91 committed:
사용되지 않는 DiscountList를 삭제하라

더 이상 사용되지 않는 DiscountList를 삭제했습니다.

eastshine-high reviewed (approved):
고생하셨습니다^^

hannut91 committed:
할인 목록을 조회할 때 페이지네이션을 추가하라

지금은 모든 할인 목록을 가져오고 있습니다. 할인이 점점 많아지면, 서버에
부하가 생기므로, 일정한 만큼만 조회하도록 허용해야 합니다. 따라서
페이지네이션을 추가하여, 일부만 조회할 수 있도록 할인목록에 페이지네이션
을 추가했습니다.

`/discounts?page=0&size=10`

See also
  - https://docs.spring.io/spring-data/jpa/docs/2.5.5/reference/html/#repositories.core-concepts
  - https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/PagingAndSortingRepository.html

hannut91 commented on src/main/java/com/kurly/promotion/ui/DiscountController.java:
```diff
@@ -39,8 +41,12 @@
-    public List<DiscountData> list() {
-        List<Discount> discountList = discountService.getDiscounts();
+    public List<DiscountData> list(
+            @RequestParam(defaultValue = "0") int page,
```
page는 0부터 시작합니다.
